Feature social say #522

jeremylan opened the pull request:
#522 
Created entries in the config file `pocs.yaml`, handling code in `core.py` and social sinks classes for Twitter and Slack.

jamessynge commented:
Thanks Jeremy. I'll take a look this weekend... I've got a very busy week in-progress now.

jamessynge commented:
I'd prefer to not have this integrated into core.py, but rather as a small binary (e.g. POCS/script/say_to_twitter.py), which takes command line args or can read from a yaml file as you've done, to configure the zeromq reading and the twitter writing. This makes it easier to test, and it doesn't further complicate the already complicated core.py?

Thoughts?

wtgee commented:
Sorry, I've been traveling to and talking at a conference so haven't looked
at this used but would tend to agree that this feature shouldn't be in
core.py. I'll try to take a look later tonight but it's long days here.

Cheers,
Wilfred

On Sun, Jun 10, 2018, 12:40 James Synge <notifications@github.com> wrote:

> I'd prefer to not have this integrated into core.py, but rather as a small
> binary (e.g. POCS/script/say_to_twitter.py), which takes command line args
> or can read from a yaml file as you've done, to configure the zeromq
> reading and the twitter writing. This makes it easier to test, and it
> doesn't further complicate the already complicated core.py?
>
> Thoughts?
>
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub
> <https://github.com/panoptes/POCS/pull/524#issuecomment-396067324>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEUUJ8FKnCFX-bu8nKUro7Gg-Hn4uBpks5t7VoYgaJpZM4UdgpN>
> .
>

wtgee commented:
Also, on quick glance, can you remove the camelCase? We're trying to
maintain a consistency with pep8.

Thanks!

On Sun, Jun 10, 2018, 14:07 Wilfred Tyler Gee <wtylergee@gmail.com> wrote:

> Sorry, I've been traveling to and talking at a conference so haven't
> looked at this used but would tend to agree that this feature shouldn't be
> in core.py. I'll try to take a look later tonight but it's long days here.
>
> Cheers,
> Wilfred
>
> On Sun, Jun 10, 2018, 12:40 James Synge <notifications@github.com> wrote:
>
>> I'd prefer to not have this integrated into core.py, but rather as a
>> small binary (e.g. POCS/script/say_to_twitter.py), which takes command line
>> args or can read from a yaml file as you've done, to configure the zeromq
>> reading and the twitter writing. This makes it easier to test, and it
>> doesn't further complicate the already complicated core.py?
>>
>> Thoughts?
>>
>> —
>> You are receiving this because your review was requested.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/panoptes/POCS/pull/524#issuecomment-396067324>, or mute
>> the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEUUJ8FKnCFX-bu8nKUro7Gg-Hn4uBpks5t7VoYgaJpZM4UdgpN>
>> .
>>
>

jeremylan commented:
Hi thanks for the comments.

Yes I can certainly have it a separate script. I left in core as there was already an asynchronous handler for commands so I reused the similar code and made it totally configurable from the config. Also the slack output was in core.

If I have it on a separate script, how should it get invoked? I.e. as a totally separate process invoked from command line or invoked by `pocs_shell`?

Regarding the camel case no probs, as it is passing the unit test for consistency I thought it was ok.

If you could give me guidance on the invocation question above I'll modify the PR accordingly.

Cheers

wtgee commented:
So I think we want to make this a separate script in the `scripts` dir  (or perhaps just part of `scripts/start_messaging_hub.py`? @jamessynge?).

jeremylan commented:
Thanks for the comments @wtgee 

No problem about the response time, I'm also working at this when I can.
I will start cleaning up `core.py` and also implement the changes you commented on, no issue!
It's not a critical feature so no rush :)

Regarding the camel casing, I assume you mean the class names `SocialTwitter` and `SocialSlack` right? I named that way as I saw `PanMessaging`. I will rename them `social_twitter` and `social_slack` ok?

I'll wait for @jamessynge guidance, but I think I got the scripts architecture a little better now, there is a `start_xxxx` script in the `startup` directory and the real code (which I can bring over from `core.py`) in the `run_xxxx` script.

So I could really have `start_social_messaging.sh` and `run_social_messaging.py`.
Can I leave the config stuff in `pocs.yaml` or should I create a separate one?

Cheers and enjoy the conference.

wtgee commented:
> Regarding the camel casing, I assume you mean the variables in code like socialConfig right? I'll change the names.

Yup, the class names are fine. Relevant [pep8 here](https://www.python.org/dev/peps/pep-0008/#naming-conventions).

> I'll wait for @jamessynge guidance, but I think I got the scripts architecture a little better now, there is a start_xxxx script in the startup directory and the real code (which I can bring over from core.py) in the run_xxxx script.
> So I could really have start_social_messaging.sh and run_social_messaging.py.

Yeah, sounds good about the above.

> Can I leave the config stuff in pocs.yaml or should I create a separate one?
 
Let's keep in `pocs.yaml` for now but keep it commented out

Thanks!

jeremylan commented:
Amended PR based on feedback @wtgee @jamessynge. Sorry it took a while.
I restored `core.py` to the version before my initial commit.
Hope this new changes are OK.
Thanks

jamessynge commented on pocs/utils/social_twitter.py:
```diff
@@ -0,0 +1,30 @@
+
+            self.api = tweepy.API(auth)
+            self.output_timestamp = output_timestamp
+        except tweepy.TweepError as e:
```
Does this error mean that the object is unusable? If so, it would probably be best to raise the exception again.

jamessynge commented on conf_files/pocs.yaml:
```diff
@@ -52,4 +52,15 @@ messaging:
+#         access_token_secret: [your_access_token_secret] 
+#     slack:
+#         webhook_url: [your_webhook_url]
+#         output_timestamp: False
```
Should `output_timestamp` also appear in the twitter config?

jeremylan replied:
No that would not be allowed either. 
The rules are fuzzy but they are getting stricter. I'm not super familiar with Twitter but there is some info [here](https://twittercommunity.com/t/duplicate-tweets/13264) and [here](https://blog.smarterqueue.com/faq/march-2018-twitter-duplication-policy-is-changing/)

jamessynge replied:
Thanks for the clarification.

jamessynge commented on pocs/utils/social_slack.py:
```diff
@@ -0,0 +1,25 @@
+
+class SocialSlack(object):
+
+    """Messaging class to output to Slack
```
Nit: please use punctuation. And in this case I'd end the docstring on that first line, e.g.:

`"""Messaging class to output to Slack."""`

jamessynge commented on pocs/utils/social_slack.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    logger = get_root_logger()
+
+    def __init__(self, web_hook, output_timestamp):
```
Did you want to address this?

jeremylan replied:
This was actually changed to False by default

jamessynge commented on pocs/utils/social_twitter.py:
```diff
@@ -0,0 +1,30 @@
+    """
+    logger = get_root_logger()
+
+    def __init__(self, consumer_key, consumer_secret, access_token, access_token_secret, output_timestamp):
```
I wonder whether this class should just get a dict (from the yaml reader) representing the config inside the twitter block of the config file. That would allow the caller to be relatively independent of the set of configs needed by one of these classes.

jeremylan replied:
That is certainly possible, however the style of coding in the other modules I saw is to handle the config in the 'main' module rather than the utils module. I think that is how PanMessaging works for example. 

I thought passing the parameters rather than a dictionary was a little clearer as it's a sort of DI of the stuff this class needs

jeremylan replied:
OK I'll look into this thanks

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,167 @@
+        parser.print_help()
+        sys.exit(1)
+
+    # Initialise social sinks to None
```
English, rather than American, spelling of initialize. Just curious if you are from outside of the US.

jeremylan replied:
Yes I am Sydney based. I can change this if it's a problem

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,167 @@
+    if args.from_config:
+        config = load_config(config_files=['pocs'])
+        
+        if 'social_accounts' in config:
```
Alternately, could do something like:

```
social_config = config.get('social_accounts', {})
twitter_config = social_config.get('twitter')
if twitter_config:
    try:
        social_twitter = SocialTwitter(twitter_config)
    except SocialException as e:
        log...
```

dict.get(key, default_value=None) defaults to returning None, so you don't have to worry about testing if the key is present. And you can provide a default value other than None.

jeremylan replied:
Ok that's great. I'm not familiar with Python hence I copied this style from somewhere in `core.py`. I can change to this as I also find it cleaner thanks.

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,167 @@
+            if social_config is not None and 'twitter' in social_config:
+                twitter_config = social_config['twitter']
+                if twitter_config is not None and \
+                        'consumer_key' in twitter_config and \
```
If you got with my suggestion earlier, this validation will move into the twitter class. And you could of course use a factory method rather than the constructor to do your validation, if that made things cleaner.

jeremylan replied:
See my comment above about handling config in main module

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,167 @@
+                    social_slack = SocialSlack(slack_config['webhook_url'],
+                                                    output_timestamp)
+        else:
+            arg_error('social_accounts setting not defined in config.')
```
I wouldn't emit an error for this, only if the config is actually bad.

jeremylan replied:
ok

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,167 @@
+        else:
+            arg_error('social_accounts setting not defined in config.')
+
+    if not social_twitter and not social_slack:
```
Let's treat this as normal. That way our startup scripts can always include executing this script, and it will just exit if there are no social accounts specified, with a non-error message such as "No social messaging sinks defined, exiting."

jeremylan replied:
Ok sure

jamessynge commented:
Thanks very much for the adjustments, I think this is getting very close.

jamessynge commented:
Oh, and can you think how to test this?

jamessynge commented on pocs/utils/social_slack.py:
```diff
@@ -0,0 +1,25 @@
+import requests
```
I'm not familiar with the requests library. Where does that come from?

jeremylan replied:
This was in @wtgee code he originally used to output to Slack.
See the original code [here](https://github.com/panoptes/POCS/blob/may-run/pocs/core.py#L185).
I just reused it.

wtgee replied:
It's not standard python but it's more straight-forward to use than `urllib` or `urllib2`. I was actually surprised we weren't already using it somewhere. It's a popular library so I don't have any issues including it.

jamessynge replied:
We don't have this in requirements.txt. @wtgee do you know how we are getting it installed?

wtgee replied:
Let's put it in the requirements. We hadn't used this before so wasn't there.

jeremylan commented:
Hi @jamessynge thanks for the review. Please see my individual  comments inline. 
I'll address your points with another commit as soon as I get some time.
Thanks

jeremylan commented:
Hi @jamessynge, @wtgee 

I just pushed another commit addressing (hopefully) all the points we discussed.

Regarding testing, I'm not too sure how to write an unit test for this. I guess we could mock config values and send some tweets/slacks from the unit test, however test accounts would be needed...

jamessynge commented:
Thanks, I'll take a look soon. Regarding testing, I found these links with
a quick search:

https://bobbelderbos.com/2016/06/unittest-with-python/

https://github.com/python-glasgow/pythonglasgow/blob/master/tests/test_twitter.py


On Mon, Jun 18, 2018, 10:07 PM Luca <notifications@github.com> wrote:

> Hi @jamessynge <https://github.com/jamessynge>, @wtgee
> <https://github.com/wtgee>
>
> I just pushed another commit addressing (hopefully) all the points we
> discussed.
>
> Regarding testing, I'm not too sure how to write an unit test for this. I
> guess we could mock config values and send some tweets/slacks from the unit
> test, however test accounts would be needed...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/panoptes/POCS/pull/524#issuecomment-398251349>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABYmUwN-MbrIzQJVz85erohb38MVIfTOks5t-FzbgaJpZM4UdgpN>
> .
>

On Jun 18, 2018 10:07 PM, "Luca" <notifications@github.com> wrote:

Hi @jamessynge <https://github.com/jamessynge>, @wtgee
<https://github.com/wtgee>

I just pushed another commit addressing (hopefully) all the points we
discussed.

Regarding testing, I'm not too sure how to write an unit test for this. I
guess we could mock config values and send some tweets/slacks from the unit
test, however test accounts would be needed...

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/panoptes/POCS/pull/524#issuecomment-398251349>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABYmUwN-MbrIzQJVz85erohb38MVIfTOks5t-FzbgaJpZM4UdgpN>
.

jeremylan commented:
Hi @jamessynge 

Thanks for the links about unit testing, they were useful. Python's mocking framework is very interesting.

I added the unit test module, it took me a little to finish as `pytest` does not support mock decorators and it puzzled me for a while. It all works using the `with` format.

I also made a little change in the sink classes to better handle empty parameters, and I removed inconsistent use of single and double quotes.

Thanks

jamessynge commented on pocs/tests/test_social_messaging.py:
```diff
@@ -0,0 +1,103 @@
+
+
+def test_send_message_twitter_no_timestamp(twitter_config):
+    with unittest.mock.patch.dict(twitter_config, {'output_timestamp': False}):
```
FYI, for this kind of nested `with` you can put multiple `with_items` in a single with statement. See the docs here: https://docs.python.org/3/reference/compound_stmts.html#the-with-statement

jeremylan replied:
Will change this thanks

jamessynge commented on pocs/tests/test_social_messaging.py:
```diff
@@ -0,0 +1,103 @@
+import pytest
```
Great job on the tests, thanks.

jamessynge commented on pocs/utils/social_slack.py:
```diff
@@ -0,0 +1,28 @@
+    def send_message(self, msg, timestamp):
+        try:
+            if self.output_timestamp:
+                post_msg = '{} - {}'.format(timestamp, msg)
```
FWIW, I think we might find it easier to consume these messages if the timestamp comes second, after msg. For example, on my phone I'd rather see timestamp truncated off than the msg.

jeremylan replied:
OK will do

jamessynge commented on pocs/utils/social_twitter.py:
```diff
@@ -0,0 +1,47 @@
+    def send_message(self, msg, timestamp):
+        try:
+            if self.output_timestamp:
+                retStatus = self.api.update_status('{} - {}'.format(timestamp, msg))
```
Ditto here about order.

jamessynge commented on scripts/run_social_messaging.py:
```diff
@@ -0,0 +1,149 @@
+    parser.add_argument(
+        '--from_config',
+        action='store_true',
+        help='Read social channels config from the pocs.yaml and pocs_local.yaml config files.')
```
At this point we don't really have any need for argparse, right? If you agree, I'd drop it until needed.

wtgee replied:
@jamessynge you mean because there is only one option or just because it is the default?

I'd prefer that if we parse _any_ command line options we do it with ArgParse, even if it is only one argument. It's more consistent and ArgParse looks cleaner than using `sys.argv`.

wtgee replied:
However, what we haven't done is establish conventions on naming arguments from the command line. I think the standard would be to always do hypens instead of underscores on the command line, which ArgParse automatically adjusts. So `--from-config` on the command line becomes `args.from_config`.

jeremylan replied:
So @wtgee, @jamessynge  nothing to do here then right? BTW I lifted this code from `run_messaging_hub.py`
